### PR TITLE
Changes for modelica airflow modelling

### PR DESCRIPTION
**Pull request by kldjonge**
Based on the documentation of CONTAM, new components and related baseclasses were developed for the IBPSA.Airflow.Multizone package:

### Multizone
- IBPSA.Airflow.Multizone.Q_Cdp
- IBPSA.Airflow.Multizone.F_Cdp
- IBPSA.Airflow.Multizone.FvsP
- IBPSA.Airflow.Multizone.QvsP
- IBPSA.Airflow.Multizone.Powerlaw_1Datapoint
- IBPSA.Airflow.Multizone.Powerlaw_2Datapoints

### Multizone.BaseClasses
- IBPSA.Airflow.Multizone.BaseClasses.PartialOneWayFlowelement
- **(adapted)** IBPSA.Airflow.Multizone.BaseClasses.PowerLawResistance 
- **(adapted)** IBPSA.Airflow.Multizone.BaseClasses.PowerLawResistance_mflow 
- IBPSA.Airflow.Multizone.BaseClasses.WindpressureProfile()
- IBPSA.Airflow.Multizone.BaseClasses.flowElementData()

Additionally two components were added to the IBPSA.Fluid package namely:

### IBPSA.Fluid
- IBPSA.Fluid.Sources.Outside_CpData
- IBPSA.Fluid.Movers.PrescribedFlow

### CODE REVIEW

File: IBPSA/Airflow/Multizone/BaseClasses/PowerLawResistance.mo [hunk 1]
```diff
@@ -96,119 +30,30 @@ equation
...
-<code>m</code> a parameter.
-For turbulent flow, set <code>m=1/2</code> and
-for laminar flow, set <code>m=1</code>.
-</p>
-<p>
-The model is used as a base for the interzonal air flow models.
-</p>
+<p>This model describes the mass flow rate and pressure difference relation in the form </p>
+<p><span style=\"font-family: Courier New;\">V_flow = k * dp^m,</span></p>
+<p>where <span style=\"font-family: Courier New;\">k</span>  and <span style=\"font-family: Courier New;\">m</span> are parameters. For turbulent flow, set <span style=\"font-family: Courier New;\">m=1/2</span> and for laminar flow, set <span style=\"font-family: Courier New;\">m=1</span>. </p>
```
Comment (Mathadon) [inline IBPSA/Airflow/Multizone/BaseClasses/PowerLawResistance.mo:38]: @kldjonge whenever you make changes to the html documentation and revision history using the dymola 'info editor', dymola reformats the html code, which leads to very long lines as you can see here. We try to format everything as in the original format to ensure readability. Therefore please revert these changes and use the dymola 'info source' tab to edit the documentation instead of the 'info editor'.

File: IBPSA/Airflow/Multizone/BaseClasses/PowerLawResistance.mo [hunk 2]
```diff
@@ -96,119 +30,30 @@ equation
...
-<li>December 6, 2011 by Michael Wetter:<br/>
-       Removed <code>fixed=false</code> attribute of protected parameter
-       <code>k</code>.
-</li>
-<li>July 20, 2010 by Michael Wetter:<br/>
-       Migrated model to Modelica 3.1 and integrated it into the Buildings library.
-</li>
-<li>February 4, 2005 by Michael Wetter:<br/>
-       Released first version.
+<li>June 26, 2020, by Klaas De Jonge (UGent):<br/>Implementation using IBPSA.Airflow.Multizone.BaseClasses.PartialOneWayFlowelement</li>
```
Comment (Mathadon) [inline IBPSA/Airflow/Multizone/BaseClasses/PowerLawResistance.mo:43]: Same comment here with respect to html formatting.

File: IBPSA/Airflow/Multizone/BaseClasses/WindpressureProfile.mo [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+algorithm
+
+  // Convert  from radians to degrees
+  beta:=u*(180/Modelica.Constants.pi);
+
+  // Change sign to positive
+  aR := if beta < 0 then -beta else beta;
+
+  // Constrain to [0...2*pi]
+  if aR > 360 then
```
Comment (Mathadon) [inline IBPSA/Airflow/Multizone/BaseClasses/WindpressureProfile.mo:34]: You can use the function `mod()` here if you want. It leads to shorter code.

File: IBPSA/Airflow/Multizone/BaseClasses/WindpressureProfile.mo [hunk 4]
```diff
@@ -0,0 +1,63 @@
+within IBPSA.Airflow.Multizone.BaseClasses;
+function WindpressureProfile
```
Comment (Mathadon) [inline IBPSA/Airflow/Multizone/BaseClasses/WindpressureProfile.mo:2]: I think that the according to the naming conventions we'd use the name `WindPressureProfile` instead of `WindpressureProfile`.

File: IBPSA/Airflow/Multizone/BaseClasses/flowElementData.mo [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        y1=yd[i],
+        y2=yd[i + 1],
+        y1d=d[i],
+        y2d=d[i + 1]);
+  end if;
+
+  annotation (
+    Icon(coordinateSystem(preserveAspectRatio=false)),
+    Diagram(coordinateSystem(preserveAspectRatio=false)),
+    experiment(StopTime=8, __Dymola_Algorithm="Dassl"),
```
Comment (Mathadon) [inline IBPSA/Airflow/Multizone/BaseClasses/flowElementData.mo:51]: I don't think that you can simulate a function so these simulation annotations can be removed?

File: IBPSA/Airflow/Multizone/EffectiveAirLeakageArea.mo [hunk 6]
```diff
@@ -14,6 +14,7 @@ model EffectiveAirLeakageArea "Effective air leakage area"
     annotation (Dialog(group="Rating conditions"));
 
   parameter Modelica.SIunits.Area L(min=0) "Effective leakage area";
+  Modelica.SIunits.Velocity v(nominal=1) "Average velocity";
```
Comment (Mathadon) [inline IBPSA/Airflow/Multizone/EffectiveAirLeakageArea.mo:17]: You added a real variable but it was already used on line 19, so it existed already?

File: IBPSA/Airflow/Multizone/Orifice.mo [hunk 7]
```diff
@@ -4,10 +4,10 @@ model Orifice "Orifice"
     m=0.5,
     k=CD*A*sqrt(2.0/rho_default));
 
-  parameter Modelica.SIunits.Area A "Area of orifice"
-    annotation (Dialog(group="Orifice characteristics"));
+  parameter Modelica.SIunits.Area A "Area of orifice" annotation (Dialog(group="Orifice characteristics"));
   parameter Real CD=0.65 "Discharge coefficient"
     annotation (Dialog(group="Orifice characteristics"));
+  Modelica.SIunits.Velocity v(nominal=1) "Average velocity";
```
Comment (Mathadon) [inline IBPSA/Airflow/Multizone/Orifice.mo:10]: Same remark as before: v already exists?

File: IBPSA/Airflow/Multizone/package.order [hunk 8]
```diff
@@ -1,10 +1,16 @@
 UsersGuide
+MediumColumn
```
Comment (Mathadon) [inline IBPSA/Airflow/Multizone/package.order:2]: The model order was revised, meaning they are no longer alphabetical. This will cause unit tests to fail. The package order should be alphabetical (except for some package names like 'Examples', 'Validation', etc.)

File: IBPSA/BoundaryConditions/Types.mo [hunk 9]
```diff
@@ -9,8 +9,7 @@ package Types "Package with type definitions"
 <p>
 Enumeration to define the data source used in the weather data reader.
 </p>
-</html>", revisions=
-"<html>
+</html>", revisions="<html>
```
Comment (Mathadon) [inline IBPSA/BoundaryConditions/Types.mo:12]: Try to avoid these model changes that do not actually contain any significant code changes. Especially if they make the formatting worse. You can use 'git checkout master --  IBPSA/BoundaryConditions/Types.mo' and commit to revert this change.

File: IBPSA/Fluid/Movers/PrescribedFlow.mo [hunk 10]
```diff
@@ -0,0 +1,78 @@
+within IBPSA.Fluid.Movers;
+model PrescribedFlow
+  "Component for idealised, prescribed massflow or volume flow"
+  extends Airflow.Multizone.BaseClasses.PartialOneWayFlowelement;
+
+  parameter Boolean FlowOpt = true
```
Comment (Mathadon) [inline IBPSA/Fluid/Movers/PrescribedFlow.mo:6]: Variable names that start with a word should start with a non-capital letter.

File: IBPSA/Fluid/Movers/PrescribedFlow.mo [hunk 11]
```diff
@@ -0,0 +1,78 @@
+within IBPSA.Fluid.Movers;
```
Comment (Mathadon) [inline IBPSA/Fluid/Movers/PrescribedFlow.mo:1]: I think this belongs in the Airflow package since it's only used there?

File: IBPSA/package.mo [hunk 12]
```diff
@@ -486,7 +486,9 @@ annotation (
 version="3.0.0",
 versionDate="2018-09-28",
 dateModified = "2018-09-28",
-uses(Modelica(version="3.2.3")),
+uses(Modelica(version="3.2.3"),
+    CONTAM_Modelica(version="2"),
```
Comment (Mathadon) [inline IBPSA/package.mo:490]: please remove the dependencies on CONTAM_Modelica and IDEAS.

File: IBPSA/Airflow/Multizone/EffectiveAirLeakageArea.mo (hunk 6 above)
Comment (kldjonge) [reply IBPSA/Airflow/Multizone/EffectiveAirLeakageArea.mo:17]: The average velocity variable is specific to the type of flow element and was moved to a higher level.  Only EffectiveAirLeakageArea and Orifice use this variable.

File: IBPSA/Airflow/Multizone/Orifice.mo (hunk 7 above)
Comment (kldjonge) [reply IBPSA/Airflow/Multizone/Orifice.mo:10]: The average velocity variable is specific to the type of flow element and was moved to a higher level.  Only EffectiveAirLeakageArea and Orifice use this variable.

File: IBPSA/Fluid/Movers/PrescribedFlow.mo (hunk 11 above)
Comment (kldjonge) [reply IBPSA/Fluid/Movers/PrescribedFlow.mo:1]: I wasn't sure yet because looking at the component  itself it is not actually restricted to airflow only.

File: IBPSA/Airflow/Multizone/BaseClasses/WindpressureProfile.mo (hunk 4 above)
Comment (kldjonge) [reply IBPSA/Airflow/Multizone/BaseClasses/WindpressureProfile.mo:2]: As it is function it should start without capital, changed it to windPressureProfile

File: IBPSA/package.mo (hunk 12 above)
Comment (kldjonge) [reply IBPSA/package.mo:490]: is there a quick check to make sure all dependencies are gone? Do you know if Dymola updates this code accordingly if all dependencies are gone?

Comment (Mathadon) [reply IBPSA/package.mo:490]: I'm pretty sure that dymola does not update it. I would do a search using grep (linux) on the word 'CONTAM' in the library directory to figure out whether or not the library is still referenced.

File: IBPSA/Airflow/Multizone/BaseClasses/WindpressureProfile.mo (hunk 4 above)
Comment (Mathadon) [reply IBPSA/Airflow/Multizone/BaseClasses/WindpressureProfile.mo]: correct
